### ISSUE TITLE
Servers as peers to each other?

### ISSUE THREAD

**Issue by bradisbell** [share/sharedb#158]
Is it possible to use ShareDB in a mode where multiple servers are peers to each other?  Could I have a pool of 4 or 5 servers connected to each other, all updating each other just like clients do?

I'd imagine I would have to implement the network transport part, and would plan to do so over multicast.

**Comment by curran** [comment]
Yes, this is possible. This is exactly the scenario that [sharedb-redis-pubsub](https://github.com/share/sharedb-redis-pubsub) was created for. This add-on gives you horizontal scalability.

**Comment by bradisbell** [comment]
I'd like to skip the Redis dependency.  I'd like to have a setup where the servers communicate with each other directly, so that there isn't a single point of failure, nor a cluster of Redis servers that need to be managed.  Are you saying that if I simply implement what the sharedb-redis-pubsub implements but with my own transport, that is going to work?

**Comment by curran** [comment]
I think so, but @nateps could give a definitive answer.

Also just to clarify - you only need a single Redis instance, possibly on a single machine, that manages the pub/sub. But as you mention that gives a single point of failure.

**Comment by sferoze** [comment]
@curran are you currently using this redis driver with sharedb?

I have it implemented by passing the sharedb-redis driver to options.pubsub. And the driver successfully connects to the redis database, but no keys or data is ever written to redis. In the logs it just shows that the server connects. 

Is there any other options I have to set to get this to work?

**Comment by curran** [comment]
@sferoze I am indeed.

That behavior is to be expected - the [sharedb-redis-pubsub](https://github.com/share/sharedb-redis-pubsub/blob/master/index.js)  library never persists anything to Redis. It only uses the publish/subscribe mechanism built into Redis (see the [source code](https://github.com/share/sharedb-redis-pubsub/blob/master/index.js#L63)).

If you want to verify that Redis is being used, maybe try to find a way to monitor the "publish" calls. I'm not sure how to do this.

**Comment by zag2art** [comment]
It's not hard to implement your own package using any technology you want. for example we made this for a project - [https://github.com/dmapper/sharedb-wsbus-pubsub](https://github.com/dmapper/sharedb-wsbus-pubsub). It uses our custom websocket server. Redis was not approved for the project. But this case you will have to provide reliability by yourself.

**Comment by zag2art** [comment]
BTW, in my opinion redis is extremely reliable. Working over three years with derby, racer, sharejs/db I can say that I've not seen any bugs or failures because of redis.

**Comment by curran** [comment]
@nateps @josephg @rkstedman [issue tending note] I think this issue can be closed, because the requested functionality is possible via:

 * [sharedb-redis-pubsub](https://github.com/share/sharedb-redis-pubsub)
 * [sharedb-wsbus-pubsub](https://github.com/dmapper/sharedb-wsbus-pubsub)

Related to **Link to sharedb-wsbus-pubsub** #166
